File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/ImageExamples.java
package questpdf.docexamples;

import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSize;
import com.questpdf.helpers.Placeholders;
import com.questpdf.infrastructure.DocumentSettings;
import com.questpdf.infrastructure.Image;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;

// NOT PORTED: image-dynamic.webp — depends on .NET SkiaSharp (SKBitmap/SKCanvas synthesize the dynamic image).
public class ImageExamples extends DocExample {

    @Test
    public void example() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(400f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .grid(grid -> {
                            grid.columns(2);
                            grid.spacing(10f);

                            grid.item(2).text("My photo gallery:").bold();

                            grid.item().image(resource("Photos/photo-gallery-1.jpg"));
                            grid.item().image(resource("Photos/photo-gallery-2.jpg"));
                            grid.item().image(resource("Photos/photo-gallery-3.jpg"));
                            grid.item().image(resource("Photos/photo-gallery-4.jpg"));
                        });
                });
            })
            .generateImages(index -> output("image-example.webp"), settings);
    }

    @Test
    public void imageScaling() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(1000f, 1500f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.item().paddingBottom(5f).text("FitWidth").bold();
                            column.item()
                                .width(200f)
                                .height(150f)
                                .border(4f)
                                .borderColor(Colors.Red.getMedium())
                                .image(resource("Photos/photo.jpg"))
                                .fitWidth();

                            column.item().height(15f);

                            column.item().paddingBottom(5f).text("FitHeight").bold();
                            column.item()
                                .width(200f)
                                .height(100f)
                                .border(4f)
                                .borderColor(Colors.Red.getMedium())
                                .image(resource("Photos/photo.jpg"))
                                .fitHeight();

                            column.item().height(15f);

                            column.item().paddingBottom(5f).text("FitArea 1").bold();
                            column.item()
                                .width(200f)
                                .height(100f)
                                .border(4f)
                                .borderColor(Colors.Red.getMedium())
                                .image(resource("Photos/photo.jpg"))
                                .fitArea();

                            column.item().height(15f);

                            column.item().paddingBottom(5f).text("FitArea 2").bold();
                            column.item()
                                .width(200f)
                                .height(150f)
                                .border(4f)
                                .borderColor(Colors.Red.getMedium())
                                .image(resource("Photos/photo.jpg"))
                                .fitArea();

                            column.item().height(15f);

                            column.item().paddingBottom(5f).text("FitUnproportionally").bold();
                            column.item()
                                .width(200f)
                                .height(50f)
                                .border(4f)
                                .borderColor(Colors.Red.getMedium())
                                .image(resource("Photos/photo.jpg"))
                                .fitUnproportionally();
                        });
                });
            })
            .generateImages(index -> output("image-scaling.webp"), settings);
    }

    @Test
    public void dpiSetting() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(400f, 1000f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(10f);

                            // lower raster dpi = lower resolution, pixelation
                            column.item()
                                .image(resource("Photos/photo.jpg"))
                                .withRasterDpi(16);

                            // higher raster dpi = higher resolution
                            column.item()
                                .image(resource("Photos/photo.jpg"))
                                .withRasterDpi(288);
                        });
                });
            })
            .generateImages(index -> output("image-dpi.webp"), settings);
    }

    @Test
    public void compressionSetting() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(400f, 1000f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(10f);

                            // low quality = smaller output file
                            column.item()
                                .image(resource("Photos/photo.jpg"))
                                .withCompressionQuality(ImageCompressionQuality.VeryLow);

                            // high quality / fidelity = larger output file
                            column.item()
                                .image(resource("Photos/photo.jpg"))
                                .withCompressionQuality(ImageCompressionQuality.VeryHigh);
                        });
                });
            })
            .generateImages(index -> output("image-compression.webp"), settings);
    }

    @Test
    public void globalSettings() {
        var settings = new DocumentSettings();

        // default: ImageCompressionQuality.High;
        settings.setImageCompressionQuality(ImageCompressionQuality.Medium);

        // default: 288
        settings.setImageRasterDpi(14);

        Document
            .create(document -> {
                document.page(page -> {
                    page.content().image(resource("Photos/photo.jpg"));
                });
            })
            .withSettings(settings)
            .generatePdf(output("image-global-settings.pdf"));
    }

    @Test
    public void sharedImages() {
        var image = Image.fromFile(resource("checkbox.png"));

        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(350f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(15f);

                            for (var i = 0; i < 5; i++) {
                                column.item().row(row -> {
                                    row.autoItem().width(28f).image(image);
                                    row.relativeItem().paddingLeft(8f).alignMiddle().text(Placeholders.label());
                                });
                            }
                        });
                });
            })
            .generateImages(index -> output("image-shared.webp"), settings);
    }

    @Test
    public void svgSupport() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.continuousSize(250f);
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    final String svgContent;

                    try {
                        svgContent = Files.readString(Path.of(resource("pdf-icon.svg")));
                    } catch (IOException exception) {
                        throw new UncheckedIOException(exception);
                    }

                    page.content()
                        .column(column -> {
                            column.item().text("The classic PDF icon looks like this:").bold();
                            column.item().height(15f);
                            column.item().svg(svgContent);
                        });
                });
            })
            .generateImages(index -> output("image-svg.webp"), settings);
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/ImageExamples.kt
package questpdf.docexamples

import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSize
import com.questpdf.helpers.Placeholders
import com.questpdf.infrastructure.DocumentSettings
import com.questpdf.infrastructure.Image
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import org.junit.jupiter.api.Test
import java.io.File

// NOT PORTED: image-dynamic.webp — depends on .NET SkiaSharp (SKBitmap/SKCanvas synthesize the dynamic image).
class ImageExamples : DocExample() {

    @Test
    fun example() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(400f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .grid {
                            columns(2)
                            spacing(10f)

                            item(2).text("My photo gallery:").bold()

                            item().image(resource("Photos/photo-gallery-1.jpg"))
                            item().image(resource("Photos/photo-gallery-2.jpg"))
                            item().image(resource("Photos/photo-gallery-3.jpg"))
                            item().image(resource("Photos/photo-gallery-4.jpg"))
                        }
                }
            }
            .generateImages({ output("image-example.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun imageScaling() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(1000f, 1500f))
                    margin(25f)

                    content()
                        .column {
                            item().paddingBottom(5f).text("FitWidth").bold()
                            item()
                                .width(200f)
                                .height(150f)
                                .border(4f)
                                .borderColor(Colors.Red.Medium)
                                .image(resource("Photos/photo.jpg"))
                                .fitWidth()

                            item().height(15f)

                            item().paddingBottom(5f).text("FitHeight").bold()
                            item()
                                .width(200f)
                                .height(100f)
                                .border(4f)
                                .borderColor(Colors.Red.Medium)
                                .image(resource("Photos/photo.jpg"))
                                .fitHeight()

                            item().height(15f)

                            item().paddingBottom(5f).text("FitArea 1").bold()
                            item()
                                .width(200f)
                                .height(100f)
                                .border(4f)
                                .borderColor(Colors.Red.Medium)
                                .image(resource("Photos/photo.jpg"))
                                .fitArea()

                            item().height(15f)

                            item().paddingBottom(5f).text("FitArea 2").bold()
                            item()
                                .width(200f)
                                .height(150f)
                                .border(4f)
                                .borderColor(Colors.Red.Medium)
                                .image(resource("Photos/photo.jpg"))
                                .fitArea()

                            item().height(15f)

                            item().paddingBottom(5f).text("FitUnproportionally").bold()
                            item()
                                .width(200f)
                                .height(50f)
                                .border(4f)
                                .borderColor(Colors.Red.Medium)
                                .image(resource("Photos/photo.jpg"))
                                .fitUnproportionally()
                        }
                }
            }
            .generateImages({ output("image-scaling.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun dpiSetting() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(400f, 1000f))
                    margin(25f)

                    content()
                        .column {
                            spacing(10f)

                            // lower raster dpi = lower resolution, pixelation
                            item()
                                .image(resource("Photos/photo.jpg"))
                                .withRasterDpi(16)

                            // higher raster dpi = higher resolution
                            item()
                                .image(resource("Photos/photo.jpg"))
                                .withRasterDpi(288)
                        }
                }
            }
            .generateImages({ output("image-dpi.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun compressionSetting() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(400f, 1000f))
                    margin(25f)

                    content()
                        .column {
                            spacing(10f)

                            // low quality = smaller output file
                            item()
                                .image(resource("Photos/photo.jpg"))
                                .withCompressionQuality(ImageCompressionQuality.VeryLow)

                            // high quality / fidelity = larger output file
                            item()
                                .image(resource("Photos/photo.jpg"))
                                .withCompressionQuality(ImageCompressionQuality.VeryHigh)
                        }
                }
            }
            .generateImages({ output("image-compression.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun globalSettings() {
        Document
            .create {
                page {
                    content().image(resource("Photos/photo.jpg"))
                }
            }
            .withSettings(DocumentSettings().apply {
                // default: ImageCompressionQuality.High;
                imageCompressionQuality = ImageCompressionQuality.Medium

                // default: 288
                imageRasterDpi = 14
            })
            .generatePdf(output("image-global-settings.pdf"))
    }

    @Test
    fun sharedImages() {
        val image = Image.fromFile(resource("checkbox.png"))

        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(350f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(15f)

                            repeat(5) {
                                item().row {
                                    autoItem().width(28f).image(image)
                                    relativeItem().paddingLeft(8f).alignMiddle().text(Placeholders.label())
                                }
                            }
                        }
                }
            }
            .generateImages({ output("image-shared.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun svgSupport() {
        Document
            .create {
                page {
                    continuousSize(250f)
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    val svgContent = File(resource("pdf-icon.svg")).readText()

                    content()
                        .column {
                            item().text("The classic PDF icon looks like this:").bold()
                            item().height(15f)
                            item().svg(svgContent)
                        }
                }
            }
            .generateImages({ output("image-svg.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }
}
